### PR TITLE
perf: Improve `normalizeMethod`

### PR DESCRIPTION
**Pull request by tsctx**
## Status

- [ ] I have read and agreed to the [Developer's Certificate of Origin][cert]
- [ ] Tested
- [ ] Benchmarked (**optional**)
- [ ] Documented
- [ ] Review ready
- [ ] In review
- [ ] Merge ready

[cert]: https://github.com/nodejs/undici/blob/main/CONTRIBUTING.md

### CODE REVIEW

**Review by mcollina — APPROVED** [review]
lgtm

**Comment by ronag** [comment]
Benchmark?

File: lib/fetch/util.js
Comment (KhafraDev) [inline lib/fetch/util.js:717]: Can you add a test that uses fetch that attempts to access something on Object.prototype?

**Review by KhafraDev — CHANGES_REQUESTED** [review]

File: lib/fetch/util.js
Comment (tsctx) [reply lib/fetch/util.js:717]: Unit tests are added with 41c1e743ad55a63fc87d8d9c0873a9cbb7caa059

**Review by KhafraDev — APPROVED** [review]
Changes look fine but lets wait for CI to be fixed before landing.

**Comment by tsctx** [comment]
Hi @ronag, 
See this benchmark.
```js
// bench.js
const { Request } = require("./lib/fetch/request");
const benchmark = require("benchmark");

const suite = new benchmark.Suite();

const input = "https://example.com/post";

const init = {
  method: "POST",
};

suite.add("new Request()", () => {
  new Request(input, init);
});

suite.on("cycle", function (event) {
  console.log(String(event.target));
});

// run async
new Promise((resolve) => setTimeout(resolve, 7000)).then(() =>
  suite.run({ async: true })
);
```

- *main*
```
new Request() x 42,952 ops/sec ±5.67% (81 runs sampled)
```
- *this patch*
```
new Request() x 43,595 ops/sec ±3.76% (81 runs sampled)
```

**Review by ronag — APPROVED** [review]

**Comment by tsctx** [comment]
Listing `['delete', 'DELETE', 'get', 'GET', 'head', 'HEAD', 'options', 'OPTIONS', 'post', 'POST', 'put', 'PUT']` below is 2.5 times faster than the current regex implementation.
But otherwise it is about 1.6 times slower.
Most of the time, I use the ones on this list, so it's not a problem.

**Comment by ronag** [comment]
Patch is missing

**Comment by tsctx** [comment]
Furthermore, if you want to pursue performance, you can use cache in Map.
This is the fastest way.
```js
const normalizeMethodMap = new Map(
  Object.entries({
    delete: "DELETE",
    DELETE: "DELETE",
    get: "GET",
    GET: "GET",
    head: "HEAD",
    HEAD: "HEAD",
    options: "OPTIONS",
    OPTIONS: "OPTIONS",
    post: "POST",
    POST: "POST",
    put: "PUT",
    PUT: "PUT",
  })
);

let methodMapCache = new Map(normalizeMethodMap);

function impl_map_cache(method) {
  const cache = methodMapCache.get(method);
  if (cache !== undefined) return cache;
  if (methodMapCache.size > 4096) methodMapCache = new Map(normalizeMethodMap);
  const c = normalizeMethodMap.get(method.toLowerCase()) ?? method;
  methodMapCache.set(method, c);
  return c;
}
```

**Comment by tsctx** [comment]
@ronag 
Adding a PATCH is nice, but you need two objects.

**Comment by ronag** [comment]
> @ronag Adding a PATCH is nice, but you need two objects.

Sorry? I don't understand.

**Review by ronag — APPROVED** [review]

**Comment by tsctx** [comment]
@ronag 
In short, it comes down to this.
```js
const normalizeMethodRecordBase = {
  delete: "DELETE",
  DELETE: "DELETE",
  get: "GET",
  GET: "GET",
  head: "HEAD",
  HEAD: "HEAD",
  options: "OPTIONS",
  OPTIONS: "OPTIONS",
  post: "POST",
  POST: "POST",
  put: "PUT",
  PUT: "PUT",
};

const normalizeMethodRecord = {
  ...normalizeMethodRecordBase,
  patch: "patch",
  PATCH: "PATCH",
};

Object.setPrototypeOf(normalizeMethodRecordBase, null);

Object.setPrototypeOf(normalizeMethodRecord, null);

function impl_object(method) {
  return (
    normalizeMethodRecord[method] ??
    // Here, don't use normalizeMethodRecord, otherwise `PATCH` becomes `patch`.
    normalizeMethodRecordBase[method.toLowerCase()] ??
    method
  );
}
```

**Comment by ronag** [comment]
I don't understand. Why not just add it to the first object?

**Comment by ronag** [comment]
```
const normalizeMethodRecordBase = {
  delete: "DELETE",
  DELETE: "DELETE",
  get: "GET",
  GET: "GET",
  head: "HEAD",
  HEAD: "HEAD",
  options: "OPTIONS",
  OPTIONS: "OPTIONS",
  post: "POST",
  POST: "POST",
  put: "PUT",
  PUT: "PUT",
  patch: "PATCH",
  PATCH: "PATCH",
};
```

**Comment by tsctx** [comment]
The specification reads as follows.
```
To normalize a [method](https://fetch.spec.whatwg.org/#concept-method), if it is a [byte-case-insensitive](https://infra.spec.whatwg.org/#byte-case-insensitive) match for `DELETE`, `GET`, `HEAD`, `OPTIONS`, `POST`, or `PUT`, [byte-uppercase](https://infra.spec.whatwg.org/#byte-uppercase) it.
```
It does not say that `patch` is to be `PATCH`.

**Comment by tsctx** [comment]
An example specification states.
Using `patch` is highly likely to result in a `405 Method Not Allowed`. `PATCH` is much more likely to succeed.

**Pull request merged**
